modules/core/shared/src/main/scala/gem/util/Timestamp.scala

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gem.util
+
+import cats._
+import cats.implicits._
+import cats.effect.IO
+
+import java.time.{ Instant, ZonedDateTime }
+import java.time.ZoneOffset.UTC
+import java.time.temporal.ChronoUnit.MICROS
+
+
+/** Timestamp wraps a `java.util.Instant` that is truncated to microsecond
+  * resolution.  This allows Timestamps to roundtrip to/from the database
+  * where only microsecond resolution is supported.  In addition the min
+  * and max instants that are supported are determined by the postgres limit for
+  * timestamps.
+  *
+  * @param toInstant
+  */
+final class Timestamp private(val toInstant: Instant) extends AnyVal {
+
+  /** Gets the number of seconds from the Java epoch of 1970-01-01T00:00:00Z. */
+  def epochSecond: Long =
+    toInstant.getEpochSecond
+
+  /** Gets the number of microseconds after the start of the second returned
+    * by `epochSecond`.
+    */
+  def µs: Long =
+    toInstant.getNano / 1000L
+
+  /** Converts this instant to the number of milliseconds from the epoch of
+    * 1970-01-01T00:00:00Z.
+    */
+  def toEpochMilli: Long =
+    toInstant.toEpochMilli
+
+  /** Creates an updated instance of Timestamp by applying the given
+    * function to its wrapped Instant.  The computed Instant is truncated to
+    * microsecond precision and validated.
+    */
+  private def mod(f: Instant => Instant): Option[Timestamp] =
+    Timestamp.fromInstant(f(toInstant))
+
+  def plusMillis(millisToAdd: Long): Option[Timestamp] =
+    mod(_.plusMillis(millisToAdd))
+
+  def plusMicros(microsToAdd: Long): Option[Timestamp] =
+    mod(_.plusNanos(microsToAdd * 1000))
+
+  def plusSeconds(secondsToAdd: Long): Option[Timestamp] =
+    mod(_.plusSeconds(secondsToAdd))
+
+  override def toString: String =
+    toInstant.toString
+}
+
+object Timestamp {
+
+  /** Minimum time that can be stored in a postgres timestamp. */
+  val Min: Timestamp =
+    truncate(ZonedDateTime.of( -4712, 1, 1, 0, 0, 0, 0, UTC).toInstant)
+
+  /** Maximum time that can be stored in a postgres timestamp. */
+  val Max: Timestamp =
+    truncate(ZonedDateTime.of(294275, 12, 31, 23, 59, 59, 999999000, UTC).toInstant)
```

Comment:
I don't think this is the exact upper limit, but the crucial point is that there _is a limit_ and it is close to this value.